src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+    do
+    {
+        ReturnErrorOnFailure(group.Load(mStorage));
+        group.Delete(mStorage);
+        group.group = group.next;
+    } while (kUndefinedGroupId != group.group);
+
+    // Remove endpoint
+
+    if (kInvalidEndpointId == prev_eid)
+    {
+        // First endpoint, update fabric info
+        fabric.first_endpoint = endpoint.next;
+        fabric.Save(mStorage);
+    }
+    else
+    {
+        // Mid endpoint, update previous endpoint's info
+        EndpointData prev_endpoint(fabric_index, prev_eid);
+        prev_endpoint.Load(mStorage);
+        prev_endpoint.next = endpoint.next;
+        prev_endpoint.Save(mStorage);
+    }
+
+    return endpoint.Delete(mStorage);
+}
+
+GroupDataProvider::GroupMappingIterator * GroupDataProviderImpl::IterateGroupMappings(chip::FabricIndex fabric_index,
+                                                                                      EndpointId endpoint_id)
+{
+    VerifyOrReturnError(mInitialized, nullptr);
+    return mEndpointIterators.CreateObject(*this, fabric_index, endpoint_id);
+}
+
+GroupDataProviderImpl::GroupMappingIteratorImpl::GroupMappingIteratorImpl(GroupDataProviderImpl & provider,
+                                                                          chip::FabricIndex fabric_index,
+                                                                          chip::EndpointId endpoint_id) :
+    mProvider(provider),
+    mFabric(fabric_index), mEndpoint(endpoint_id)
+{
+    FabricData fabric(fabric_index);
+    ReturnOnFailure(fabric.Load(provider.mStorage));
+
+    // Existing fabric
+
+    EndpointData endpoint(fabric_index, fabric.first_endpoint);
+    uint16_t count = 0;
+
+    // Loop through the fabric's endpoints
+    do
+    {
+        ReturnOnFailure(endpoint.Load(provider.mStorage));
+        if (endpoint.id == endpoint_id)
+        {
+            // Target endpoint found
+            mGroup = endpoint.first_group;
+            break;
+        }
+        endpoint.id = endpoint.next;
+    } while (++count < fabric.endpoint_count);
+}
+
+size_t GroupDataProviderImpl::GroupMappingIteratorImpl::Count()
+{
+    size_t count = 0;
+
+    GroupData group(mFabric, mEndpoint, mGroup);
+    chip::GroupId prev_gid = kUndefinedGroupId;
+
+    while ((kUndefinedGroupId != group.group) && (prev_gid != group.group))
+    {
+        if (CHIP_NO_ERROR != group.Load(mProvider.mStorage))
+        {
+            break;
+        }
+        prev_gid    = group.group;
+        group.group = group.next;
+        count++;
+    }
+    return count;
+}
+
+bool GroupDataProviderImpl::GroupMappingIteratorImpl::Next(GroupMapping & item)
+{
+    GroupData group(mFabric, mEndpoint, mGroup);
+    CHIP_ERROR err = group.Load(mProvider.mStorage);
+    if ((kUndefinedGroupId == mGroup) || (CHIP_NO_ERROR != err))
+    {
+        return false;
+    }
+    item.endpoint = mEndpoint;
+    item.group    = mGroup;
+    strncpy(item.name, group.name, GroupData::kGroupNameMax);
```

Comment:
Please enforce null-termination (reminder, given the overall comment given much above)

Reply:
Null-termination explicitly added.